proposals/0393-parameter-packs.md

```diff
@@ -560,16 +561,30 @@ func overload<T>(_: T) {}
 func overload<each T>(_: repeat each T) {}
 ```
 
-If both overloads match a given call, e.g. `overload(1)`, the call is ambiguous. Similarly, two generic functions where one accepts a non-pack variadic parameter and the other accepts a type parameter pack:
+If the parameters of the scalar overload have the same or refined requirements as the parameter pack overload, the scalar overload is considered a subtype of the parameter pack overload, because the parameters of the scalar overload can be forwarded to the parameter pack overload. Generally, if a function call successfully type checks with two different overloads, the subtype is preferred. This effectively means that scalar overloads are preferred over parameter pack overloads when the scalar requirements meet the requirements of the parameter pack:
 
 ```swift
-func overload<T>(_: T...) {}
+func overload() {}
+func overload<T>(_: T) {}
 func overload<each T>(_: repeat each T) {}
 
-overload() // ambiguity error
+overload() // calls the no-parameter overload
+
+overload(1) // calls the scalar overload
+
+overload(1, "") // calls the parameter pack overload
 ```
 
-In other words, variadic generic functions have the same ranking as other generic functions.
+The general overload subtype ranking rule applies after localized ranking, such as implicit conversions and optional promotions. That remains unchanged with this proposal. For example:
+
+```swift
+func overload<T>(_: T, _: Any) {}
+func overload<each T>(_: repeat each T) {}
+
+overload(1, "") // prefers the parameter pack overload because the scalar overload would require an existential conversion
+```
+
+This overload resolution behavior enables library authors to introduce new function overloads using parameter packs that generalize existing fixed-arity overloads while preserving the overload resolution behavior of existing code.
```

Comment:
Is it possible to spell out what would happen with somewhat more convoluted scenarios (basically, to what extent this will "just work" and where we'll end up with an unresolvable ambiguity)? For example:

```swift
func foo<each T: BinaryInteger>(_: repeat each T, str: String, _: repeat each T) { print(str) }
func foo<each U: StringProtocol>(_: repeat each U) { print("Goodbye, world!") }

foo("Hello, world!") // prints ???
```

Reply:
I tried this approach originally, but it's actually really difficult to do without surfacing a _lot_ of implementation details about how constraint system ranking works, some of which have not been formalized through Swift evolution and we would like to change (though unlikely that we _can_ change without doing so under a language mode).

The case you mentioned above would be ambiguous. There is an existing ranking rule that will consider the number of type parameters (but only ones that appear in the argument list, not the return type, for some reason 🤷🏻‍♀️), and prefer the one with fewer type parameters, but there is no ranking rule that considers the number of times a type parameter appears in the parameter list. There's also no attempt to consider a generic overload "more concrete" if a concrete type appears somewhere in the function signature, unless it falls out of the more general overload subtype comparison. The structure of solution ranking is super complicated, and none of it is introduced by this proposal (nor can it be changed by this proposal).

Reply:
I'd appreciate any thoughts you might have on that though! I want to be as descriptive as possible about what new rules are introduced by this proposal, but I also don't want to imply that some of the overload ranking behaviors can be changed, because most of the behavior falls out of what's already there.

Reply:
Totally understand. My main thought here is that the following sentences in this revision can be read to be more sweeping than what's implemented (emphasis added):

> If the parameters of the scalar overload have the same or refined requirements as the parameter pack overload, the scalar overload is considered a subtype . . . . **Generally**, if a function call successfully type checks with two different overloads, the subtype is preferred.

I wonder if it might be helpful to concretely state here (a) what _is_ assured not to be ambiguous (presumably, overload sets of the form `func f(); func f<T>(_: T); func f<T, U>(_: T, _: U)` when `func f<each T>(_: repeat each T)` is added); (b) that in more complex scenarios there will remain overloads that are considered ambiguous as a result of other ranking rules which continue to apply and won't be changed, with maybe one or two of the simplest examples that users may actually to encounter. In that way, if a more convoluted scenario is encountered, a user could read this document and say, "Ah, this ambiguity is something that _can_ happen" rather than wondering if it's an unforeseen implementation oversight.

Reply:
I added two ambiguity examples. How does it look now?

Reply:
Love it! Thanks